napari/utils/colormaps/colormap.py

```diff
@@ -294,47 +316,77 @@ def _cast_labels_to_minimum_dtype_auto(
 
-    return _modulo_plus_one(data, num_colors, dtype, background_value)
+    if colormap.use_selection:
+        return (data == colormap.selection).astype(dtype) * (
+            _zero_preserving_modulo_naive(
+                np.array([colormap.selection]), num_colors, dtype
+            )
+        )
+
+    return _zero_preserving_modulo(
+        data, num_colors, dtype, colormap.background_value
+    )
 
 
-@numba.njit(parallel=True)
-def _modulo_plus_one(
+def _zero_preserving_modulo_naive(
     values: np.ndarray, n: int, dtype: np.dtype, to_zero: int = 0
 ) -> np.ndarray:
-    """Like ``values % n + 1``, but with one specific value mapped to 0.
+    res = ((values - 1) % n + 1).astype(dtype)
+    res[values == to_zero] = 0
+    return res
 
-    This ensures (1) an output value in [0, n] (inclusive), and (2) that
-    no nonzero values in the input are zero in the output, other than the
-    ``to_zero`` value.
 
-    Parameters
-    ----------
-    values : np.ndarray
-        The dividend of the modulo operator.
-    n : int
-        The divisor.
-    dtype : np.dtype
-        The desired dtype for the output array.
-    to_zero : int, optional
-        A specific value to map to 0. (By default, 0 itself.)
+try:
+    import numba
+except ImportError:
+    _zero_preserving_modulo = _zero_preserving_modulo_naive
+else:
 
-    Returns
-    -------
-    np.ndarray
-        The result: 0 for the ``to_zero`` value, ``values % n + 1``
-        everywhere else.
-    """
-    result = np.empty_like(values, dtype=dtype)
+    @numba.njit(parallel=True)
+    def _zero_preserving_modulo(
+        values: np.ndarray, n: int, dtype: np.dtype, to_zero: int = 0
+    ) -> np.ndarray:
+        """Like ``values % n + 1``, but with one specific value mapped to 0.
 
-    for i in numba.prange(values.size):
-        if values.flat[i] == to_zero:
-            result.flat[i] = 0
-        else:
-            result.flat[i] = values.flat[i] % n + 1
+        This ensures (1) an output value in [0, n] (inclusive), and (2) that
+        no nonzero values in the input are zero in the output, other than the
+        ``to_zero`` value.
+
+        Parameters
+        ----------
+        values : np.ndarray
+            The dividend of the modulo operator.
+        n : int
+            The divisor.
+        dtype : np.dtype
+            The desired dtype for the output array.
+        to_zero : int, optional
+            A specific value to map to 0. (By default, 0 itself.)
+
+        Returns
+        -------
+        np.ndarray
+            The result: 0 for the ``to_zero`` value, ``values % n + 1``
+            everywhere else.
+        """
+        result = np.empty_like(values, dtype=dtype)
+
+        for i in numba.prange(values.size):
+            if values.flat[i] == to_zero:
+                result.flat[i] = 0
+            else:
+                result.flat[i] = (values.flat[i] - 1) % n + 1
 
-    return result
+        return result
```

Comment:
Very minor suggestion/preference — feel free to ignore it.

I kind of don't like that this function definition is nested. It makes the flow a bit trickier to understand for me. So my suggestion would be:

1. At the top of the module, do the numba import attempt:

```python
try:
    import numba
    numba_available = True
except ModuleNotFoundError:
    numba_available = False
```

2. Later, define the function normally, no jit:

```python
def _zero_preserving_modulo_numpy(
    ...

def _zero_preserving_modulo_loop(
    ...
```

3. Finally, a simple if/else:

```python
if numba_available:
    _zero_preserving_modulo = numba.njit(_zero_preserving_modulo_loop, parallel=True)
else:
    _zero_preserving_modulo = _zero_preserving_modulo_numpy
```

Reply:
The problem with this if/else is that it will raise multiple linters. But maybe I should extract all to `colormap_utils_numba.py`?

Reply:
What problems?

Reply:
To me, that is a problem with the linter — I guess there is a `# noqa` comment that we can add to that line?